contributors/design-proposals/node/pod-pid-namespace.md

```diff
@@ -1,77 +1,150 @@
 
+The shared PID functionality will be hidden behind a new feature gate in both
+the API server and the kubelet, and the `--docker-disable-shared-pid` flag will
+be removed from the kubelet.
 
-[1]: https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/
+## User Experience
 
+### Use Cases
 
-<!-- BEGIN MUNGE: GENERATED_ANALYTICS -->
-[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/docs/proposals/pod-pid-namespace.md?pixel)]()
-<!-- END MUNGE: GENERATED_ANALYTICS -->
+Sharing a PID namespace between containers in a pod is discussed in
+[#1615](https://issues.k8s.io/1615) and enables:
+
+1.  signaling between containers, which is useful for side cars (e.g. for
+    signaling a daemon process after rotating logs).
+1.  easier troubleshooting of pods.
+1.  addressing [Docker's zombie
+    problem](https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/)
+    by reaping orphaned zombies in the infra container.
+
+## Implementation
+
+### Kubernetes API Changes
+
+`v1.PodSecurityContext` gains a new field resembling the existing `hostPID`
+field:
+
+```
+type PodSecurityContext struct {
+    ...
+    // Use the host's pid namespace.
+    // Optional: Default to false.
+    // +k8s:conversion-gen=false
+    // +optional
+    HostPID bool
+    // Use a single shared PID namespace for the pod.
+    // Optional: Default to false.
+    // +k8s:conversion-gen=false
+    // +optional
+    SharedPID bool
+       ...
+```
+
+`PodSecurityContext` was chosen because it's where namespace configuration
+currently lives, and it is consistent with using isolated PID namespaces as a
+security feature. Setting both `SharedPID` and `HostPID` will cause a validation
+error.
+
+### Container Runtime Interface Changes
+
+Namespace options in the CRI are specified for both PodSandbox and Container
+creation requests. We will add a new field to NamespaceOption:
+
+```
+message NamespaceOption {
+    ...
+    // If set, use the host's PID namespace.
+    bool host_pid = 2;
+    ...
+    // If set, use the pod's shared PID namespace.
+    bool shared_pid = 4;
+}
+```
+
+The kubelet runtime manager will set `shared_pid` to
+`PodSecurityContext.SharedPID` for both the `LinuxSandboxSecurityContext` and
+`LinuxContainerSecurityContext`.
+
+### dockershim Changes
+
+The Docker runtime implements the pod sandbox as a container running the pause
+container image. The `shared_pid` of `LinuxSandboxSecurityContext` will be
+ignored by this runtime as the sandbox container will become the namespace
+that's shared with the containers in the pod.
+
+The dockershim will translate a `CreateContainerRequest` with `shared_pid` set
+to true in `LinuxContainerSecurityContext` to a call to Docker's create
+container request with `PidMode` set to the container ID of the sandbox
+container. All containers in the pod will then start in the same PID namespace
+as the sandbox container, with the `/pause` binary as PID 1.
```

Comment:
Does it make sense to allow users to use their own custom sandbox image with pid 1 capabilities?

Reply:
There's nothing preventing users to configure their runtime/shim to use a different image, if the runtime/shim supports it. I don't see we support that in kube/CRI in the near future though.

Reply:
It'd be great to support such a feature, but I think it would be complicated. That's more towards the "general purpose init" to which I alluded in Non-goals.